prow/cmd/validation-webhook/main.go

```diff
@@ -0,0 +1,140 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"net/http"
+
+	"github.com/sirupsen/logrus"
+	secretmanagerpb "google.golang.org/genproto/googleapis/cloud/secretmanager/v1"
+	"k8s.io/test-infra/experiment/clustersecretbackup/secretmanager"
+	"k8s.io/test-infra/prow/flagutil"
+	"k8s.io/test-infra/prow/logrusutil"
+)
+
+const (
+	secretID  = "ca-certification"
+	secretMapKey = "ca-secrets"
+	caCert = "ca-cert"
+	caPrivKey = "ca-priv-key"
+	caBundle = "ca-bundle"
+	latest = "projects/*/secrets/*/versions/latest"
+	gcpSecretError = "could not get secret data"
+)
+
+type ClientInterface interface {
+	CreateSecret(ctx context.Context, secretID string) (*secretmanagerpb.Secret, error)
+	AddSecretVersion(ctx context.Context, secretName string, payload []byte) error
+	GetSecretValue(ctx context.Context, secretName, versionName string) ([]byte, error)
+}
+
+
+
+func main() {
+	logrusutil.ComponentInit()
+	logrus.SetLevel(logrus.DebugLevel)
+	// GCP project in which to store secrets in Secret Manager
+	dnsNames := []string{"validation-webhook-service",
+		"validation-webhook-service.default", "validation-webhook-service.default.svc"}
+	p := flag.String("p", 
+	"colew-test", "Project ID for storing GCP Secrets")
+	e := flag.Int("expiry-years", 30, "CA certificate expiry in years")
+	dns := flagutil.NewStrings()
+	for _, dnsName := range(dnsNames) {
+		dns.Add(dnsName)
+	}
+	flag.Var(&dns, "dns", "DNS Names CA-Cert config")
+	flag.Parse()
+	exp := *e
+	projectId := *p
+	client, err := secretmanager.NewClient(projectId, false)
+	if err != nil {
+		logrus.WithError(err).Fatal("Unable to create secret manager client")
+	}
+	ctx := context.Background()
+	cert, privKey, err := getGCPSecrets(client, ctx, exp, dns)
+	if err != nil && err.Error() == gcpSecretError {
+		logrus.WithError(err)
+		cert, privKey, err = createGCPSecret(client, ctx, exp, dns)
+		if err != nil {
+			logrus.WithError(err).Fatal("Unable to create ca certificate")
+		}
+	} else if err.Error() != gcpSecretError {
+		logrus.WithError(err).Fatal("Unable to get GCP Secret")
+	}
+	err = isCertValid(cert)
+	if err != nil {
+		cert, privKey, err = createGCPSecret(client, ctx, exp, dns)
+		if err != nil {
+			logrus.WithError(err).Fatal("Unable to create GCP Secret")
+		}
+	} 
+	http.HandleFunc("/validate", serveValidate)
+	logrus.Info("Listening on port 8008...")
+	logrus.Fatal(http.ListenAndServeTLS(":8008", cert, privKey, nil))
+}
+
+func getGCPSecrets(client ClientInterface, ctx context.Context, expiry int, dns flagutil.Strings) (string, string, error) {
+	secretsMap := make(map[string]string)
+	data, err := client.GetSecretValue(ctx, secretID, latest)
+	if err != nil {
+		return "", "", fmt.Errorf(gcpSecretError)
+	}
```

Comment:
This ignores the message contents in `err` and returns a potentially inaccurate error. `GetSecretValue` could fail due to auth issues or network problems, we don't know that the error indicates the secret doesn't exist.

Also if you want to distinguish an error you shouldn't match against error text like you do on line 74. It is preferable to use type checking instead to see if the error is of a specific type.

Reply:
type error interface {
	Error() [string](https://pkg.go.dev/builtin#string)
}
this is what the error interface provides

Reply:
what exactly do you mean?